### ISSUE TITLE
`bad-file-permissions` (S103) has false negatives and false positives

### ISSUE THREAD

**Issue by dscorbett** [astral-sh/ruff#18863]
### Summary

[`bad-file-permissions` (S103)](https://docs.astral.sh/ruff/rules/bad-file-permissions/) has false negatives and false positives.

[The upstream rule checks `0o33`](https://github.com/PyCQA/bandit/blob/1.8.5/bandit/plugins/general_bad_file_permissions.py#L63-L69) but Ruff’s S103 only checks `0o12`.
```console
$ cat >s103_1.py <<'# EOF'
import os
os.chmod("/etc/secrets.txt", 0o21)
# EOF

$ flake8 ./s103_1.py
./s103_1.py:2:1: S103 Chmod setting a permissive mask 0o21 on file (/etc/secrets.txt).

$ ruff --isolated check s103_1.py --select S103
All checks passed!
```

Only the bits of `0o33` matter for the permissiveness check. S103 should not be suppressed when those bits can statically be determined in the result of a bitwise OR, even when the full result can’t.
```python
import os
def f(path, mode):
    os.chmod(path, mode | 0o777)
```

There is a false positive when the result of a bitwise AND can statically be determined to be valid, even when a subexpression has an invalid value.
```python
$ cat >s103_2.py <<'# EOF'
import os
os.chmod("/etc/secrets.txt", 0o777777 & 0o700)
os.chmod("/etc/secrets.txt", 0o777777 & 0o777)  # This is a true positive for the wrong reason.
# EOF

$ ruff --isolated check s103_2.py --select S103 --output-format concise -q
s103_2.py:2:30: S103 `os.chmod` setting an invalid mask on file or directory
s103_2.py:3:30: S103 `os.chmod` setting an invalid mask on file or directory
```

### Version

ruff 0.12.0 (87f0feb21 2025-06-17)

**Comment by MichaReiser** [comment]
That Ruff only checks Group execute and world write, but not group write or world execute, has been the case since the rule was added (https://github.com/astral-sh/ruff/pull/1636/). 

The upstream rule distinguishes between group write/execute (medium) and world (write/execute) but that's not what's implemented in Ruff. https://github.com/PyCQA/bandit/pull/570/files


The rule does have special handling for bitwise operands but it only supports values up to `u16` as the left/right side:

https://github.com/astral-sh/ruff/blob/daa385c1a991cc87c77642eccdcc2a888d3586a0/crates/ruff_linter/src/rules/flake8_bandit/rules/bad_file_permissions.rs#L171-L176

**Comment by Peiffap** [comment]
I'd be open to taking a stab at this.

@MichaReiser, do you have additional context on why the Ruff rule differs from the upstream rule? Would aligning them be desirable?
As for the `u16` limit, is there a specific reason to keep it? For reference, this limitation came about in #7584 to get rid of the `num-bigint` dependency/improve performance.

**Comment by MichaReiser** [comment]
> @MichaReiser, do you have additional context on why the Ruff rule differs from the upstream rule? Would aligning them be desirable?

No, not more than what you can find from looking at the git history. Aligning with upstream is probably desired but we should double check that the change is in line with the rule's intent (and we probably need to gate it behind preview mode)

> As for the u16 limit, is there a specific reason to keep it? For reference, this limitation came about in https://github.com/astral-sh/ruff/pull/7584 to get rid of the num-bigint dependency/improve performance.

I don't think so. We shouldn't re-introduce `num-bigint`, it's probably enough to change the check to use `usize` instead (or `u64`)

**Comment by getehen** [comment]
Hi @Peiffap, has this been fixed ? If no, I can take a look

**Comment by Peiffap** [comment]
Hi @getehen ! I'm not sure, feel free to take it if not.

**Comment by getehen** [comment]
Thanks, I'll give it a try

**Comment by getehen** [comment]
@MichaReiser I am adding u64 support, and since we are checking the type of the value, what would you think of checking only the 9 first bits are used ? Something like 0o1000 would thus be invalid

**Comment by amyreese** [comment]
Consider what we do in [RUF064](https://docs.astral.sh/ruff/rules/non-octal-permissions/). Unix permissions allow for twelve (including setuid/setgid/sticky) bits, so a value like `0o1000` is technically a valid unix permission bitflag.

**Comment by yash27-lab** [comment]
I took a pass on this and prepared a patch in my fork:

- Branch: [`codex/s103-bad-file-permissions`](https://github.com/yash27-lab/ruff/tree/codex/s103-bad-file-permissions)
- Compare view: [astral-sh/ruff main...yash27-lab:codex/s103-bad-file-permissions](https://github.com/astral-sh/ruff/compare/main...yash27-lab:ruff:codex/s103-bad-file-permissions)
- Commit: `108f7cd`

What I changed:
- Aligned S103 dangerous-bit check with Bandit semantics (`0o33`): `S_IWOTH`, `S_IWGRP`, `S_IXGRP`, `S_IXOTH`.
- Switched mask analysis to `u64`.
- Added partial bitwise analysis for `|`, `&`, `^` so we can still reason about dangerous bits when full values are not fully known.
- Kept invalid-mask detection for known final values above `0o7777` (and kept `0o1000` valid, per the RUF064 note).

Issue repros now behave as expected:
- `os.chmod(..., 0o21)` -> flagged
- `os.chmod(..., mode | 0o777)` -> flagged
- `os.chmod(..., 0o777777 & 0o700)` -> not flagged
- `os.chmod(..., 0o777777 & 0o777)` -> flagged

Testing:
- Updated the `S103.py` fixture with repro + edge cases.
- Added two exhaustive tests:
  - Literal matrix over `0o0000..0o7777` (4096 cases)
  - Bitwise matrix (`|`, `&`, `^`) over `0o000..0o177` (49,152 expressions)
- Ran:
  - `cargo test -p ruff_linter s103_exhaustive_literal_matrix -- --nocapture`
  - `cargo test -p ruff_linter s103_exhaustive_bitwise_matrix -- --nocapture`
  - `cargo test -p ruff_linter rules::flake8_bandit::tests:: -- --nocapture`
- Result: all passing.

If this direction looks good, I can open the upstream PR right away.

**Comment by yash27-lab** [comment]
Fixes #18863

## Summary

- Align `S103` dangerous-bit detection with Bandit semantics:
  - `S_IWOTH` (`0o2`)
  - `S_IWGRP` (`0o20`)
  - `S_IXGRP` (`0o10`)
  - `S_IXOTH` (`0o1`)
- Switch mask evaluation to `u64`.
- Add partial bitwise analysis for `|`, `&`, and `^` so dangerous bits are detected even when full values are partially unknown.
- Preserve permission validity behavior:
  - keep `0o1000` valid
  - emit `Invalid` only for known final values above `0o7777`
- Expand `S103` fixture coverage with the reported edge cases.
- Add exhaustive tests for `S103`:
  - literal matrix over `0o0000..0o7777`
  - bitwise matrix over `0o000..0o177` for `|`, `&`, `^`

## Test Plan

- `cargo test -p ruff_linter s103_exhaustive_literal_matrix -- --nocapture`
- `cargo test -p ruff_linter s103_exhaustive_bitwise_matrix -- --nocapture`
- `cargo test -p ruff_linter rules::flake8_bandit::tests:: -- --nocapture`
